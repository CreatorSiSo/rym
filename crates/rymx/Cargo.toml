[package]
name = "rymx"
version = "0.0.3"
edition = "2021"
default-run = "rymx"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
anyhow = "1.0"
clap = { version = "4.4", features = ["cargo"] }
insta = { version = "1.31.0", features = ["glob"] }
logos = "0.13.0"
owo-colors = "3.5"
rustyline = "12"

[features]
expand = []

[profile.dev.package.insta]
opt-level = 3

[profile.dev.package.similar]
opt-level = 3
