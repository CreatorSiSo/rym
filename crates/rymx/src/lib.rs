use std::sync::mpsc::Sender;

mod ast;
mod compile;
mod error;
mod interpret;
mod parse;
mod span;
pub mod std_lib;
mod tokenize;

pub use error::AriadneEmitter;
pub use interpret::Env;
pub use tokenize::tokenizer;

use error::{Diagnostic, Level, SourceId};
use interpret::{Interpret, Value};
use span::Span;
use tokenize::Token;

pub fn interpret(env: &mut Env, ast: impl Interpret) -> Value {
    // TODO does this make sense
    // Ignoring control flow
    let result = ast.eval(env).inner();

    // let env_state: String = env
    //     .variables()
    //     .into_iter()
    //     .fold((0, String::new()), |(ident, mut accum), scope| {
    //         for (name, (kind, value)) in scope {
    //             writeln!(accum, "{}{kind} {name} = {}", "  ".repeat(ident), value)
    //                 .expect("Internal Error: Unable to write into String");
    //         }
    //         (ident + 1, accum)
    //     })
    //     .1;
    // emitter.push_result(&env_state);
    // emitter.send(t);

    result
}

pub fn compile_module(
    emitter: Sender<Diagnostic>,
    src: &str,
    src_id: SourceId,
) -> Option<ast::Module> {
    let tokens: Vec<(Token, Span)> = tokenize(emitter.clone(), src)?
        .into_iter()
        .map(|(token, span)| (token, span.with_id(src_id)))
        .collect();

    let module = match parse::parse_file(&tokens, src, src_id) {
        Ok(module) => module,
        Err(diagnostics) => {
            for diagnostic in diagnostics {
                emitter.send(diagnostic).unwrap();
            }
            return None;
        }
    };
    emitter
        .send(Diagnostic::new(Level::Debug, format!("{module:#?}\n")))
        .unwrap();

    // TODO Name resolution
    // TODO Typechecking
    // TODO Const evaluation
    // TODO Generate intermediate representation

    Some(module)
}

// TODO take a module (for name lookup and so on) as input
pub fn compile_stmt(emitter: Sender<Diagnostic>, src: &str, src_id: SourceId) -> Option<ast::Stmt> {
    let tokens: Vec<(Token, Span)> = tokenize(emitter.clone(), src)?
        .into_iter()
        .map(|(token, span)| (token, span.with_id(src_id)))
        .collect();

    let expr = match parse::parse_stmt(&tokens, src, src_id) {
        Ok(expr) => expr,
        Err(diagnostics) => {
            for diagnostic in diagnostics {
                emitter.send(diagnostic).unwrap();
            }
            return None;
        }
    };
    emitter
        .send(Diagnostic::new(Level::Debug, format!("{expr:#?}\n")))
        .unwrap();

    Some(expr)
}

fn tokenize(emitter: Sender<Diagnostic>, src: &str) -> Option<Vec<(Token, Span)>> {
    let results: Vec<_> = tokenizer(src).collect();

    let tokens_string = results.iter().fold(String::new(), |accum, (result, span)| {
        let (token, span) = match result {
            Ok(token) => (format!("{token:?}"), span),
            Err(_) => ("Error".into(), span),
        };
        accum + &format!("{token} [{}]\n", span.src(src).escape_debug())
    });
    emitter
        .send(
            Diagnostic::new(Level::Debug, "Finished tokenizing").with_child(
                vec![],
                Level::Debug,
                tokens_string,
            ),
        )
        .unwrap();

    let mut tokens = vec![];
    for (result, span) in results {
        let Ok(token) = result else {
            emitter
                .send(Diagnostic::spanned(
                    span,
                    Level::Error,
                    format!("Invalid character [{}]", span.src(src)),
                ))
                .unwrap();
            return None;
        };
        match token {
            Token::DocComment | Token::Comment | Token::VSpace | Token::HSpace => continue,
            _ => tokens.push((token, span)),
        }
    }
    Some(tokens)
}
