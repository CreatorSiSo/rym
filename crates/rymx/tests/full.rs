use insta::assert_snapshot;
use rymx::{std_lib, AriadneEmitter, Env};

#[test]
fn main() {
    insta::glob!("**/*.rym", |path| {
        let src = std::fs::read_to_string(path).unwrap();
        let mut out: Vec<u8> = vec![];
        let writer = strip_ansi_escapes::Writer::new(&mut out);
        let (sender, mut emitter) = AriadneEmitter::new(writer);
        let src_id = emitter.source_map.add(path.to_string_lossy(), &src);

        std::thread::spawn(move || {
            let mut env = Env::new(sender.clone())
                .with_constants(std_lib::CONSTANTS.into_iter().chain(std_lib::OTHER));
            let module = rymx::compile_module(sender, &src, src_id)?;
            rymx::interpret(&mut env, module);
            Some(())
        });

        emitter.emit_all_blocking();
        assert_snapshot!(String::from_utf8(out).unwrap());
    })
}
