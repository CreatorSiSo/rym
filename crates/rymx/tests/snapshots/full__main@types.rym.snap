---
source: crates/rymx/tests/full.rs
expression: "String::from_utf8(out).unwrap()"
input_file: crates/rymx/tests/types.rym
---
Debug: Finished tokenizing
Const [const]
HSpace [ ]
Ident [u1_min]
Colon [:]
HSpace [ ]
Ident [u1]
HSpace [ ]
Assign [=]
HSpace [ ]
Int [0]
Semi [;]
VSpace [\n]
Const [const]
HSpace [ ]
Ident [u1_max]
Colon [:]
HSpace [ ]
Ident [u1]
HSpace [ ]
Assign [=]
HSpace [ ]
Int [1]
Semi [;]
VSpace [\n\n]
Const [const]
HSpace [ ]
Ident [u8_min]
Colon [:]
HSpace [ ]
Ident [u8]
HSpace [ ]
Assign [=]
HSpace [ ]
Int [0]
Semi [;]
VSpace [\n]
Const [const]
HSpace [ ]
Ident [u8_max]
Colon [:]
HSpace [ ]
Ident [u8]
HSpace [ ]
Assign [=]
HSpace [ ]
Int [255]
Semi [;]
VSpace [\n\n]
Comment [// TODO ...]
VSpace [\n\n]
Fn [fn]
HSpace [ ]
Ident [array_test]
ParenOpen [(]
ParenClose [)]
HSpace [ ]
ThickArrow [=>]
HSpace [ ]
BraceOpen [{]
VSpace [\n]
Comment [// test \"arrays\" {]
VSpace [\n]
HSpace [    ]
Let [let]
HSpace [ ]
Ident [zeroed]
Colon [:]
HSpace [ ]
BracketOpen [[]
Ident [_]
BracketClose []]
Ident [u8]
HSpace [ ]
Assign [=]
HSpace [ ]
BracketOpen [[]
Int [0]
Semi [;]
HSpace [ ]
Int [512]
BracketClose []]
Semi [;]
VSpace [\n]
HSpace [    ]
Comment [// TODO allow type expressions in place of normal expressions?]
VSpace [\n]
HSpace [    ]
Comment [// assert_eq(type_of(zeroed), [512]u8);]
VSpace [\n\n]
HSpace [    ]
Let [let]
HSpace [ ]
Ident [initialized]
Colon [:]
HSpace [ ]
BracketOpen [[]
Ident [_]
BracketClose []]
Ident [u8]
HSpace [ ]
Assign [=]
HSpace [ ]
BracketOpen [[]
Int [1]
Comma [,]
HSpace [ ]
Int [0]
Comma [,]
HSpace [ ]
Int [0]
Comma [,]
HSpace [ ]
Int [255]
BracketClose []]
Semi [;]
VSpace [\n]
HSpace [    ]
Comment [// assert_eq(type_of(initialized), [4]u8);]
VSpace [\n]
HSpace [    ]
Ident [assert_eq]
ParenOpen [(]
Ident [initialized]
Dot [.]
BracketOpen [[]
Int [0]
BracketClose []]
Comma [,]
HSpace [ ]
Int [1]
ParenClose [)]
Semi [;]
VSpace [\n]
HSpace [    ]
Ident [assert_eq]
ParenOpen [(]
Ident [initialized]
Dot [.]
BracketOpen [[]
Int [1]
BracketClose []]
Comma [,]
HSpace [ ]
Int [0]
ParenClose [)]
Semi [;]
VSpace [\n]
HSpace [    ]
Ident [assert_eq]
ParenOpen [(]
Ident [initialized]
Dot [.]
BracketOpen [[]
Int [2]
BracketClose []]
Comma [,]
HSpace [ ]
Int [0]
ParenClose [)]
Semi [;]
VSpace [\n]
HSpace [    ]
Ident [assert_eq]
ParenOpen [(]
Ident [initialized]
Dot [.]
BracketOpen [[]
Int [3]
BracketClose []]
Comma [,]
HSpace [ ]
Int [255]
ParenClose [)]
Semi [;]
VSpace [\n\n]
HSpace [    ]
Let [let]
HSpace [ ]
Mut [mut]
HSpace [ ]
Ident [mutable]
Colon [:]
HSpace [ ]
BracketOpen [[]
Ident [_]
BracketClose []]
Ident [u8]
HSpace [ ]
Assign [=]
HSpace [ ]
BracketOpen [[]
Int [0]
Semi [;]
HSpace [ ]
Int [5]
BracketClose []]
Semi [;]
VSpace [\n]
HSpace [    ]
Ident [initialized]
Dot [.]
BracketOpen [[]
Int [0]
BracketClose []]
HSpace [ ]
Assign [=]
HSpace [ ]
Int [1]
Semi [;]
VSpace [\n]
HSpace [    ]
Ident [initialized]
Dot [.]
BracketOpen [[]
Int [1]
BracketClose []]
HSpace [ ]
Assign [=]
HSpace [ ]
Int [2]
Semi [;]
VSpace [\n]
HSpace [    ]
Ident [initialized]
Dot [.]
BracketOpen [[]
Int [2]
BracketClose []]
HSpace [ ]
Assign [=]
HSpace [ ]
Int [3]
Semi [;]
VSpace [\n]
HSpace [    ]
Ident [initialized]
Dot [.]
BracketOpen [[]
Int [3]
BracketClose []]
HSpace [ ]
Assign [=]
HSpace [ ]
Int [4]
Semi [;]
VSpace [\n]
HSpace [    ]
Ident [initialized]
Dot [.]
BracketOpen [[]
Int [4]
BracketClose []]
HSpace [ ]
Assign [=]
HSpace [ ]
Int [5]
Semi [;]
VSpace [\n]
HSpace [    ]
Ident [assert_eq]
ParenOpen [(]
Ident [mutable]
Comma [,]
HSpace [ ]
BracketOpen [[]
Int [1]
Comma [,]
HSpace [ ]
Int [2]
Comma [,]
HSpace [ ]
Int [3]
Comma [,]
HSpace [ ]
Int [4]
Comma [,]
HSpace [ ]
Int [5]
BracketClose []]
ParenClose [)]
Semi [;]
VSpace [\n]
BraceClose [}]
VSpace [\n\n]
Comment [// TODO Slices]
VSpace [\n]

Debug: Module {
    name: "",
    constants: [
        (
            "u1_min",
            Path(
                Path(["u1"]),
            ),
            Literal(Int: 0),
        ),
        (
            "u1_max",
            Path(
                Path(["u1"]),
            ),
            Literal(Int: 1),
        ),
        (
            "u8_min",
            Path(
                Path(["u8"]),
            ),
            Literal(Int: 0),
        ),
        (
            "u8_max",
            Path(
                Path(["u8"]),
            ),
            Literal(Int: 255),
        ),
        (
            "array_test",
            Unkown,
            Function {
                params: [],
                named_params: {},
                return_type: Unit,
                body: Block(
                    [
                        Variable(
                            Let,
                            "zeroed",
                            Array(
                                Path(
                                    Path(["_"]),
                                ),
                                Path(
                                    Path(["u8"]),
                                ),
                            ),
                            ArrayWithRepeat(
                                Literal(Int: 0),
                                Literal(Int: 512),
                            ),
                        ),
                        Variable(
                            Let,
                            "initialized",
                            Array(
                                Path(
                                    Path(["_"]),
                                ),
                                Path(
                                    Path(["u8"]),
                                ),
                            ),
                            Array([Literal(Int: 1), Literal(Int: 0), Literal(Int: 0), Literal(Int: 255)]),
                        ),
                        Expr(
                            Call(
                                Ident("assert_eq"),
                                [
                                    Subscript(
                                        Ident("initialized"),
                                        Literal(Int: 0),
                                    ),
                                    Literal(Int: 1),
                                ],
                            ),
                        ),
                        Expr(
                            Call(
                                Ident("assert_eq"),
                                [
                                    Subscript(
                                        Ident("initialized"),
                                        Literal(Int: 1),
                                    ),
                                    Literal(Int: 0),
                                ],
                            ),
                        ),
                        Expr(
                            Call(
                                Ident("assert_eq"),
                                [
                                    Subscript(
                                        Ident("initialized"),
                                        Literal(Int: 2),
                                    ),
                                    Literal(Int: 0),
                                ],
                            ),
                        ),
                        Expr(
                            Call(
                                Ident("assert_eq"),
                                [
                                    Subscript(
                                        Ident("initialized"),
                                        Literal(Int: 3),
                                    ),
                                    Literal(Int: 255),
                                ],
                            ),
                        ),
                        Variable(
                            LetMut,
                            "mutable",
                            Array(
                                Path(
                                    Path(["_"]),
                                ),
                                Path(
                                    Path(["u8"]),
                                ),
                            ),
                            ArrayWithRepeat(
                                Literal(Int: 0),
                                Literal(Int: 5),
                            ),
                        ),
                        Expr(
                            Assign(
                                Subscript(
                                    Ident("initialized"),
                                    Literal(Int: 0),
                                ),
                                Literal(Int: 1),
                            ),
                        ),
                        Expr(
                            Assign(
                                Subscript(
                                    Ident("initialized"),
                                    Literal(Int: 1),
                                ),
                                Literal(Int: 2),
                            ),
                        ),
                        Expr(
                            Assign(
                                Subscript(
                                    Ident("initialized"),
                                    Literal(Int: 2),
                                ),
                                Literal(Int: 3),
                            ),
                        ),
                        Expr(
                            Assign(
                                Subscript(
                                    Ident("initialized"),
                                    Literal(Int: 3),
                                ),
                                Literal(Int: 4),
                            ),
                        ),
                        Expr(
                            Assign(
                                Subscript(
                                    Ident("initialized"),
                                    Literal(Int: 4),
                                ),
                                Literal(Int: 5),
                            ),
                        ),
                        Expr(
                            Call(
                                Ident("assert_eq"),
                                [
                                    Ident("mutable"),
                                    Array([Literal(Int: 1), Literal(Int: 2), Literal(Int: 3), Literal(Int: 4), Literal(Int: 5)]),
                                ],
                            ),
                        ),
                    ],
                ),
            },
        ),
    ],
    types: [],
    sub_modules: [],
}



